### PR TITLE
Add `linalg.cholesky` to ChainerX

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This is work-in-progress pull request!

This PR wraps [cuSOLVER's Cholesky factorization](https://docs.nvidia.com/cuda/cusolver/index.html#cuds-lt-t-gt-potrf).

The differentiation algorithm is described in [this paper](https://arxiv.org/abs/1602.07527).

With reference to #6423 and #6764

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
GPU routine is tested against NumPy output.

TODO: 
- differentiation
- CPU routine

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 1]
```diff
@@ -93,3 +93,35 @@ def test_dot_invalid(is_module, xp, device, a_shape, b_shape, dtype):
         return xp.dot(a, b)
     else:
         return a.dot(b)
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    # Special shapes
+    chainer.testing.product({
+        'a_size': [2, ],
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Could you cover edge cases such as an empty array? And to do so, how about parameterize over the complete shape instead of only specifying a dimension? Also, could you cover invalid inputs to test that correct errors are raised?

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 2]
```diff
@@ -93,3 +93,35 @@ def test_dot_invalid(is_module, xp, device, a_shape, b_shape, dtype):
...
+))
+class TestCholesky(op_utils.NumpyOpTest):
+
+    def setup(self):
+        device = chainerx.get_default_device()
+        if device.name == 'native:0':
+            pytest.skip('CPU Cholesky is not implemented')
+
+    def generate_inputs(self):
+        print(self.a_size)
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Don't forget to delete this line :)

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 3]
```diff
@@ -93,3 +93,35 @@ def test_dot_invalid(is_module, xp, device, a_shape, b_shape, dtype):
...
+        a = numpy.random.random((self.a_size, self.a_size)).astype(self.in_dtypes)
+        # Make random matrix a symmetric positive semi-definite one
+        a = a.T.dot(a)
+        return a,
+
+    def forward_xp(self, inputs, xp):
+        a, = inputs
+        L = xp.linalg.cholesky(a)
+        L[numpy.triu_indices_from(L, k=1)] = 0.0
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Please try to fix the implementation such that `triu_indices_from` is not needed.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py (hunk 1 above)
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Empty and non-square arrays are added. Errors are raised if a matrix is not positive definite and non-square.

Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: I think it's better to separate tests that expect failures? It's mostly due to the way the test helper classes are implemented but by registering certain failures as allowed, configurations that shouldn't raise these error might do so while still passing the tests.

Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: You are right, it is safer to have them separate.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py (hunk 3 above)
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: I added [np.tri](https://docs.scipy.org/doc/numpy/reference/generated/numpy.tri.html) routine. Now it will be possible to implement `triu`/`tril` using `tri` and `where`.

**Comment by IvanYashchuk** [comment]
To support Cholesky implementation also `tri`, `tril` and `triu` implementation were added to this PR.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                uplo,
+                N,
+                out_ptr,
+                N,
+                work_space,
+                work_size,
+                devInfo);
+
+            int devInfo_h = 0;
+            CheckCudaError(cudaMemcpy(&devInfo_h, devInfo, sizeof(int), cudaMemcpyDeviceToHost));
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Could you explain why device2host is required? Wouldn't this block?

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                N,
+                out_ptr,
+                N,
+                &work_size);
+
+            // POTRF execution
+            T* work_space;
+            CheckCudaError(cudaMalloc(&work_space, work_size * sizeof(T)));
+            int *devInfo;
+            CheckCudaError(cudaMalloc(&devInfo, sizeof(int)));
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Could you use the `Allocate` function of the CUDA device wrapper similar to how we allocate workspace size in `chainerx/cuda/cuda_conv.cc` to avoid calling these CUDA functions directly.

**Comment by hvy** [comment]
Maybe you know this already but you can take a look at the CI results at the bottom. It e.g. seems like Travis is failing due to style errors.

File: chainerx_cc/chainerx/cuda/cusolver.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+#include "chainerx/error.h"
+
+namespace chainerx {
+namespace cuda {
+
+void CheckCusolverError(cusolverStatus_t status);
+
+namespace cuda_internal {
+
+class CusolverHandle {
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cusolver.h]: Yes, I agree that the correct name in this case is `CusolverDnHandle`. I'll change it.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                throw ChainerxError{"Unsuccessfull potrf (Cholesky) execution. Info = ", devInfo_h};
+            }
+        };
+
+        if (a.dtype() == Dtype::kFloat32) {
+            cholesky_impl(PrimitiveType<float>{}, cusolverDnSpotrf_bufferSize, cusolverDnSpotrf);
+        } else {
+            CHAINERX_ASSERT(a.dtype() == Dtype::kFloat64);
+            cholesky_impl(PrimitiveType<double>{}, cusolverDnDpotrf_bufferSize, cusolverDnDpotrf);
+        }
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: I can change that.
 `if else` is used here: chainerx_cc/chainerx/native/native_device/dot.cc:112

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 4 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: cuSOLVER requires `devInfo` to be in device memory. To check the value it needs to be copied back to host memory.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 5 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: I'll change it to use `Allocate` function.
I am just more familiar with the pure CUDA functions :slightly_smiling_face:

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+        if (a.ndim() != 2 || out.ndim() != 2) {
+            throw DimensionError{"ChainerX cholesky supports only 2-dimensional arrays."};
+        }
+        if (a.shape()[0] != a.shape()[1]) {
+            throw DimensionError{"Matrix is not square."};
+        }
+
+        // potrf (cholesky) stores result in-place, therefore copy ``a`` to ``out`` and then pass ``out`` to the routine
+        device.backend().CallKernel<CopyKernel>(a, out);
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: To copy either the upper or the lower `a` still a special kernel is needed.
I think `device.backend().CallKernel<CopyKernel>(Triu(a), out);` looks right.

[the remark this answers is not in the document]
Comment (hvy) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Hm, should Triu be a routine? I think the following is simpler, having Tri{u,l}, merely as a kernel.
```c++
Array out = Zeros(...);  // It's not initialized with random values/empty, but zeros. 
device.backend().CallKernel<TriuKernel>(a, out);  // Modifies only the upper half of `out`. 
```

[the remark this answers is not in the document]
Comment (hvy) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: We might want Triu as a routine in the future of course but I think for this PR, it's a bit ambitious.

[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Probably it is a bit ambitious if you say so. Of course, the better place for these routines would be in separate PR, but how to handle several simultaneous PRs that depend on each other?
I can add the kernels instead, should the code for routines be removed from here?

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 5 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Workspace is now allocated using `Empty` array.
I don't know what would be a better way than current for the `devInfo`. It needs to be in GPU memory and then copied to CPU memory to check for errors.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 8 above)
[the remark this answers is not in the document]
Comment (hvy) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Dependencies between PR should be kept as sparse as possible but if needed, such as in this case, you'll probably just have to rebase.

[the remark this answers is not in the document]
Comment (hvy) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Yes, I think that's the way to go, removing the routines and only defining the kernels needed.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 4 above)
Comment (hvy) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Are you sure `devInfo` needs to be in device memory? CuPy seems to simply pass a pointer to a `size_t` casted to a `int*`?

Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: CuPy creates an array of size 1 on the device (`dev_info`) and passes the pointer to data
https://github.com/cupy/cupy/blob/master/cupy/linalg/decomposition.py#L61
then it is transferred to CPU with `status = int(dev_info[0])`

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 5 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: `devInfo` is now allocated using `device.Allocate` and then copied to CPU with `device.MemoryCopyTo`.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 9]
```diff
@@ -227,3 +227,76 @@ def forward_xp(self, inputs, xp):
...
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    chainer.testing.product({
+        'shape': [(), (2, 3), (3, 2), (6, 6)],
+        'in_dtypes': ['float32', 'float64'],
+    })
+))
+class TestCholeskyFailing(NumpyLinalgOpTest):
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Let's create tests (that are not stochastic) for positive semi-definite inputs.

**Comment by hvy** [comment]
The implementation generally LGTM but I added some comments on the tests.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 10]
```diff
@@ -227,3 +227,76 @@ def forward_xp(self, inputs, xp):
...
+    chainer.testing.product({
+        'shape': [(1, 1), (3, 3), (6, 6)],
+        'in_dtypes': ['float32', 'float64']
+    })
+))
+class TestCholesky(NumpyLinalgOpTest):
+
+    def generate_inputs(self):
+        a = numpy.random.random(self.shape).astype(self.in_dtypes)
+        # Make random square matrix a symmetric positive semi-definite one
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Now with `a = numpy.array(a.T.dot(a)) + 1e-6*numpy.eye(*self.shape)` the input is guaranteed to be positive definite.
https://mail.python.org/pipermail/scipy-user/2007-June/012490.html

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py (hunk 9 above)
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Added a test with three different positive semidefinite matrices.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 11]
```diff
@@ -227,3 +227,101 @@ def forward_xp(self, inputs, xp):
...
+        'shape': [(1, 1), (3, 3), (6, 6)],
+        'in_dtypes': ['float32', 'float64']
+    })
+))
+class TestCholesky(NumpyLinalgOpTest):
+
+    def generate_inputs(self):
+        a = numpy.random.random(self.shape).astype(self.in_dtypes)
+        # Make random square matrix a symmetric positive definite one
+        a = numpy.array(a.T.dot(a)) + 1e-6*numpy.eye(*self.shape)
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: A nitpick on the style.
```suggestion
        a = numpy.array(a.T.dot(a)) + 1e-6 * numpy.eye(*self.shape)
```

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 12]
```diff
@@ -204,5 +245,60 @@ public:
...
+        CudaSetDeviceScope scope{device.index()};
+
+        CHAINERX_ASSERT(a.ndim() == 2);
+        CHAINERX_ASSERT(out.ndim() == 2);
+        CHAINERX_ASSERT(a.shape()[0] == a.shape()[1]);
+
+        // potrf (cholesky) stores result in-place, therefore copy ``a`` to ``out`` and then pass ``out`` to the routine
+        device.backend().CallKernel<CopyKernel>(Tril(a, 0), out);
+
+        Array out_contiguous = AsContiguous(out);
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: If `out` is not contiguous, a copy is created here, meaning that nothing is written to the argument `out`. We need a temporary out array for the computations and then write the results back to `out`, in case `out` is not contiguous.

File: chainerx/_docs/routines.py [hunk 13]
```diff
@@ -607,6 +607,21 @@ def _docs_linalg():
 .. seealso:: :func:`numpy.linalg.inv`
 """)
 
+    _docs.set_doc(
+        chainerx.linalg.cholesky,
+        """cholesky(a)
+Returns the Cholesky decomposition for the square matrix ``a``.
```
Comment (hvy) [inline chainerx/_docs/routines.py]: Not a strong opinion but how about being more explicit in what it returns, i.e. the L in L * L^{H,T}.

File: chainerx_cc/chainerx/native/native_device/linalg.cc [hunk 14]
```diff
@@ -194,5 +213,51 @@ class NativeInverseKernel : public InverseKernel {
...
+        device.CheckDevicesCompatible(a, out);
+
+        CHAINERX_ASSERT(a.ndim() == 2);
+        CHAINERX_ASSERT(out.ndim() == 2);
+        CHAINERX_ASSERT(a.shape()[0] == a.shape()[1]);
+
+        // potrf (cholesky) stores result in-place, therefore copy ``a`` to ``out`` and then pass ``out`` to the routine
+        device.backend().CallKernel<CopyKernel>(Tril(a, 0), out);
+
+        Array out_contiguous = AsContiguous(out);
```
Comment (hvy) [inline chainerx_cc/chainerx/native/native_device/linalg.cc]: Ditto.

File: chainerx/_docs/routines.py (hunk 13 above)
Comment (IvanYashchuk) [reply chainerx/_docs/routines.py]: Sure, I'll add a bit more info.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 12 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Added assert to kernels to ensure that passed  `out` array is contiguous.

File: chainerx/_docs/routines.py [hunk 15]
```diff
@@ -607,6 +607,21 @@ def _docs_linalg():
...
+
+Args:
+    a (~chainerx.ndarray): Hermitian (symmetric if all elements are real),
+        positive-definite input matrix.
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply chainerx/_docs/routines.py]: Fixed.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by IvanYashchuk** [comment]
Hm, cuSOLVER from Jenkins does not raise an error for semidefinite matrices in the tests.

**Comment by hvy** [comment]
I recall you mentioned that the behavior for semi-definite inputs is inconsistent already in NumPy, that it sometimes raises an error, and sometimes don't. It's neither well documented. How about simply skipping semi-definite inputs if they're unpredictable or UB?

**Comment by IvanYashchuk** [comment]
NumPy's behavior is inconsistent in a way that some semidefinite matrices might work. However, the ones in the test all raise an error in NumPy, LAPACK and (tested only locally) cuSOLVER 10.1

**Comment by hvy** [comment]
Memo: We should document that behavior is not consistent for cuSOLVER for semi-definite matrices. Then this PR should be more or less finished.

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 16]
```diff
@@ -325,4 +325,43 @@ Array PseudoInverse(const Array& a, float rcond) {
...
+    }
+
+    // Reference:
+    // Differentiation of the Cholesky decomposition, Iain Murray https://arxiv.org/abs/1602.07527
+    {
+        BackwardBuilder bb{"cholesky", a, out};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([a_tok = bb.RetainInput(0), out_tok = bb.RetainOutput(0), a_dtype = a.dtype(), &a_device = a.device()](
+                              BackwardContext& bctx) {
+                const Array& a = bctx.GetRetainedInput(a_tok);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: You don't have to retain `a`.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 17]
```diff
@@ -351,3 +351,97 @@ def forward_xp(self, inputs, xp):
...
+        a = numpy.array(a.T.dot(a)) + 1e-6 * numpy.eye(*self.shape)
+        return a,
+
+    def forward_xp(self, inputs, xp):
+        a, = inputs
+
+        # Input has to be symmetrized for backward test to work
+        def symmetrize(A):
+            L = xp.tril(A)
+            return (L + L.T)/2.
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: It's only the diagonal that should be halved?

Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Well, it doesn't matter much here `L + L.T` is the important operation here for getting the symmetric matrix.
Usually `(A + A.T)/2` is used for symmetrizing, I took using `tril` from [autograd tests](https://github.com/HIPS/autograd/blob/master/tests/test_linalg.py#L17).

Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Thank you for the explanation.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
I noticed that 2-dimensional, square, empty matrices (i.e. shape of `(0, 0)`) fail for native while it succeeds for cuda and NumPy. It's an edge case but could you fix it?

**Comment by IvanYashchuk** [comment]
Forward is now working with square empty matrices, but backward fails because it depends on `Inverse`, which does not yet support this kind of input. I'll fix it too.

**Comment by IvanYashchuk** [comment]
> Forward is now working with square empty matrices, but backward fails because it depends on `Inverse`, which does not yet support this kind of input. I'll fix it too.

Blocked by #7980

**Comment by hvy** [comment]
Sorry but the way in which kernels are registered has been changed in https://github.com/chainer/chainer/pull/7970. Could you make similar changes to this PR?

**Comment by hvy** [comment]
#7980 has now been merged :)

**Comment by hvy** [comment]
Could you resolve the conflicts?

**Comment by IvanYashchuk** [comment]
I've run the test for Cholesky a couple of thousand times (with `pytest test_linalg.py --count N`), caught a few backward errors (like 1 per ~3k).
I think that was due to inaccurate finite-difference (FD) derivatives, so I changed `eps` parameter to increase FD accuracy. I didn't get backward errors in my environment.

**Comment by hvy** [comment]
Jenkins, test this please.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 18]
```diff
@@ -116,13 +116,16 @@ def setup(self):
         if (device.backend.name == 'native'
                 and not chainerx.linalg._is_lapack_available()):
             pytest.skip('LAPACK is not linked to ChainerX')
-        self.check_backward_options.update({'rtol': 5e-3})
-        self.check_double_backward_options.update({'rtol': 5e-3})
+        self.check_backward_options.update({
+            'eps': 1e-5, 'rtol': 1e-4, 'atol': 1e-4})
+        self.check_double_backward_options.update({
+            'eps': 1e-5, 'rtol': 1e-4, 'atol': 1e-4})
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: We've been working on fixing flaky tests for other linalg routines. How about only modifying these parameters for the Cholesky tests?

Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: I moved these tolerance parameters to the Cholesky test.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Also ran yours tests 5000 times and they all passed so I think it's stable enough now. LGTM, let me merge this PR.
